crates/uv-fs/src/path.rs

```diff
@@ -303,6 +303,98 @@ pub fn relative_to(
     Ok(up.join(stripped))
 }
 
+/// A path that can be serialized and deserialized in a portable way by converting Windows-style
+/// backslashes to forward slashes, and using a `.` for an empty path.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PortablePath<'a>(&'a Path);
```

Comment:
I think this is mostly fine. But the big caveat here is that this implementation assumes the file path is valid UTF-8. We generally already assume this everywhere in `uv`, and undoing that assumption would be annoying (but possible). In any case, because we assume UTF-8 everywhere already, assuming it here seems fine too. But I would definitely put a mention of it in the docs. (And that somewhat makes the name "portable" a little suspicious, although I see why you chose that name.)

In terms of the actual "wrapper type" mechanics, I think the main alternative in this design space would be to define `PortablePath` as a DST like `Path` is. But doing so requires `unsafe` and IDK if it's worth doing here. The nice thing about the DST approach is that the lifetime specifier gets elevated to the borrow and removed from the type. So you'd have `&'a PortablePath` instead of `PortablePath<'a>`. That in turn makes it possible to implement certain traits like `AsRef<PortablePath>`. You could even have a `impl AsRef<PortablePath> for Path`.

The extent to which the "wrapper type" mechanics matters depends a lot on how it's intended to use this type. If it's mostly just intended to be used in a few places, then keeping its API slim and devoted to the integration points makes sense. If instead we wanted to use this everywhere we use `Path`, then we'd want something more elaborate probably.

Reply:
For some more elaborate examples that use DSTs like `Path` (and therefore require `#repr(transparent)` and unsafe code in the constructors), you can check out what we've been doing in red-knot:
- https://github.com/astral-sh/ruff/blob/main/crates/ruff_db/src/system/path.rs
- https://github.com/astral-sh/ruff/blob/main/crates/ruff_db/src/vendored/path.rs

(I've no idea if that's the way you want to go or not)

Reply:
Oh, and with regards to the UTF-8 problem that @BurntSushi mentions above -- in red-knot we make our `Path` and `PathBuf` structs wrappers around `camino::Utf8Path` and `camino::Utf8PathBuf` (respectively), rather than `std::path::Path` and `std::path::PathBuf`.

Reply:
Yeah we need to do something similar at some point.